### DISCUSSION TITLE
CLI Data Import

### DISCUSSION THREAD

**Discussion by wodka** [directus/directus#15934] [Feature Requests]
Hi, 

I'd like to propose a way to import data into a table / to ensure that some rows exist.

# Why?

with a specific schema we also need ie initial data next to it / we are also updating the schema in init containers when moving between different deployment stages.

Now it would be great to also be able to fill certain "lookpup" tables just to ensure the data within certain lookup tables.

# How?

`yarn directus data import:yml data.yml`

containing something like:

```
table_name:
  - id: 1
     other_field: 2
```

which in turn would insert id=1 if it does not exist and otherwise update it if found.

**Comment by azrikahar** [comment]
I think we might be able to use the existing import service for this as technically it's available [as a util endpoint](https://docs.directus.io/reference/system/utilities.html#import-data-from-file) for JSON or CSV files (which is also being used by import/export in the App).

There's probably be some aspects of the current logic that may need to be ironed out, such as flags for delimiters/seperators for CSV (relevant issue: #15874) and support for YAML format will be made easier with #14713, but it should somewhat be achievable with a custom CLI command using the init hook in the mean time. Especially when it comes to having a more opinionated format with `table_name`. (just to be clear not saying it shouldn't be supported, but more discussion will need to be had to make it more consistent with import/export in general)

Here's an example:

```js
const path = require('path');
const yaml = require('js-yaml');
const fse = require('fs-extra');
const tmp = require('tmp-promise');

module.exports = ({ init }, { services, getSchema, database, logger }) => {
	const { ImportService } = services;

	init('cli.after', ({ program }) => {
		async function importYmlAction({ collection, filepath }) {
			try {
				const filename = path.resolve(process.cwd(), filepath);
				const fileContent = await fse.readFile(filename, 'utf8');
				const importData = yaml.load(fileContent);

				const tmpFile = await tmp.file();
				await fse.writeJSON(tmpFile.path, importData);
				const tmpFileReadStream = fse.createReadStream(tmpFile.path);

				const schema = await getSchema();
				const importService = new ImportService({ schema, knex: database });

				await importService.importJSON(collection, tmpFileReadStream);

				await tmpFile.cleanup();
				process.exit(0);
			} catch (err) {
				logger.error(err);
				process.exit(1);
			}
		}

		program
			.command('import:yml')
			.option('--filepath <path>', `yaml file path`)
			.option('--collection <name>', `collection name`)
			.action(importYmlAction);
	});
};
```

and using a sample yaml file with an array:

```yaml
- id: 1
  body: Example body 1
- id: 2
  body: Example body 2
- body: Example body 4
```

It seems to be working as expected:

https://user-images.githubusercontent.com/42867097/195014436-86d2f594-bf38-41d5-bbe6-ac0d043b2083.mp4

**Reply by sevetseh28** [reply]
This works really well. Thanks! 
I loaded this as a `hooks` extension. Is that the right approach?

**Reply by nickrum** [reply]
Yes, that's the correct approach 👍

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
